processing/src/main/java/org/apache/druid/query/aggregation/AggregatorFactory.java

```diff
@@ -69,6 +69,24 @@ public VectorAggregator factorizeVector(VectorColumnSelectorFactory selectorFact
     throw new UOE("Aggregator[%s] cannot vectorize", getClass().getName());
   }
 
+  /**
+   * Creates an {@link Aggregator} based on the provided column selector factory.
+   * The returned value is a holder object which contains both the aggregator
+   * and its initial size in bytes. The callers can then invoke
+   * {@link Aggregator#aggregateWithSize()} to perform aggregation and get back
+   * the incremental memory required in each aggregate call. Combined with the
+   * initial size, this gives the total on-heap memory required by the aggregator.
+   *
+   * This flow does not require invoking {@link #guessAggregatorHeapFootprint(long)}
+   * which tends to over-estimate the required memory.
+   *
+   * @return AggregatorAndSize which contains the actual aggregator and its initial size.
+   */
+  public AggregatorAndSize factorizeWithSize(ColumnSelectorFactory metricFactory)
```

Comment:
same question about contract about returned sizes. Also, I wonder if there is anything we could do to make sure this method is overridden if `aggregateWithSize` is implemented, so that the initial size is not the max size...

Reply:
Updated the javadoc to advise on the required estimation.


> Also, I wonder if there is anything we could do to make sure this method is overridden if aggregateWithSize is implemented

I guess it is okay even if it isn't overridden because we would only be overestimating which would not cause failures, only somewhat poorer estimates.

Reply:
The other way around is probably more of an issue i.e. overriding `factorizeWithSize` but not `aggregateWithSize`. In such a case, `factorizeWithSize` would give a small initial size, which would never increase because `aggregateWithSize` would always return 0.

In either case, this issue is a problem if we use the new behaviour, i.e. `useMaxMemoryEstimates = false`. I guess we could address this is in a subsequent PR.